dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,313 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_CPU_H_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_CPU_H_
+
+#include "dali/core/boundary.h"
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Cyclic buffer used for storing input window for convolution.
+ *
+ * Wraps a pointer to a buffer of type T, with size equal to length * num_channels
+ * or just length if has_channels = false.
+ */
+template <typename T, bool has_channels = true>
+class CyclicPixelWrapper {
+ public:
+  CyclicPixelWrapper(T* ptr, int length, int num_channels = 1)
+      : data_(ptr),
+        start_(0),
+        end_(0),
+        elements_(0),
+        length_(length),
+        num_channels_(num_channels) {}
+
+  /**
+   * @brief Drop one pixel from the buffer, moving the start element.
+   */
+  void PopPixel() {
+    assert(elements_ > 0);
+    elements_--;
+    start_++;
+    WrapPosition(start_);
+  }
+
+  /**
+   * @brief Add a pixel, represented by contiguous `NumChannels()` elements and starting at `input`,
+   * to the buffer.
+   */
+  void PushPixel(const T* input) {
```

Comment:
Maybe Push should call pop if `elements_ == length_`?

Reply:
Traversing the axis with the window I know when to call Pop and Push, no need for additional logic.